core/src/tile/tileWorker.cpp

```diff
@@ -2,76 +2,129 @@
 
 #include "data/dataSource.h"
 #include "platform.h"
-#include "style/style.h"
 #include "tile/mapTile.h"
 #include "view/view.h"
+#include "tileManager.h"
+#include "style/style.h"
+#include "scene/scene.h"
+
+#include <algorithm>
 
-TileWorker::TileWorker() {
-    m_free = true;
-    m_aborted = false;
-    m_finished = false;
+#define WORKER_NICENESS 10
+
+TileWorker::TileWorker(TileManager& _tileManager, int _num_worker)
+    : m_tileManager(_tileManager) {
+    m_running = true;
+
+    for (int i = 0; i < _num_worker; i++) {
+        m_workers.emplace_back(&TileWorker::run, this);
+    }
 }
 
-void TileWorker::abort() {
-    m_aborted = true;
+TileWorker::~TileWorker(){
+    if (m_running) {
+        stop();
+    }
 }
 
-void TileWorker::processTileData(std::unique_ptr<TileTask> _task,
-                                 const std::vector<std::unique_ptr<Style>>& _styles,
-                                 const View& _view) {
+void TileWorker::run() {
+
+    setCurrentThreadPriority(WORKER_NICENESS);
+
+    while (true) {
+
+        std::shared_ptr<TileTask> task;
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+
+            m_condition.wait(lock, [&, this]{
+                    return !m_running || !m_queue.empty();
+                });
+
+            // Check if thread should stop
+            if (!m_running) {
+                break;
+            }
+
+            // Remove all canceled tasks
+            std::remove_if(m_queue.begin(), m_queue.end(),
+                [](const std::shared_ptr<TileTask>& a) {
+                    return a->tile->isCanceled();
+                });
 
-    m_task = std::move(_task);
-    m_free = false;
-    m_finished = false;
-    m_aborted = false;
+            if (m_queue.empty()) {
+                continue;
+            }
+
+            // Pop highest priority tile from queue
+            auto it = std::min_element(m_queue.begin(), m_queue.end(),
+                [](const std::shared_ptr<TileTask>& a, const std::shared_ptr<TileTask>& b) {
+                    if (a->tile->isVisible() != b->tile->isVisible())
+                        return a->tile->isVisible();
+
+                    return a->tile->getPriority() < b->tile->getPriority();
```

Comment:
What happens when both a and b are not visible! (Is it possible to have a tile which is not visible but still not cancelled)?